### PR TITLE
Avoid loading of a dynamic engine twice

### LINKED ISSUE #17023: Engine crashes when loading the configuration more than once
Overview:
-----------
OpenSSL's configuration file can declare an acceleration engine, which usually looks like the following:
```
[openssl_def]
engines = engine_section

[engine_section]
ibmca = ibmca_section

[ibmca_section]
# The openssl engine path for ibmca.so.
# Set the dynamic_path to where the ibmca.so engine
# resides on the system.
dynamic_path = /usr/local/lib/ibmca.so
engine_id = ibmca
init = 1
```

The flow for loading such an engine is triggered by `int_engine_configure`, and when using a dynamic path it will dynamically load the engine's `.so` file using the engine's exported `bind` function. Once bound successfully, the engine is added by name (`engine_id`) to the engine list, which can only contain a **single** engine instance per such id.

As the engine list can only contain a single instance per engine_id, all engines that I've seen to date are built as singletons, which is an understandable implementation decision considering the engine API.

The Problem:
---------------
In the past there were multiple previous bugs of engines crashing after they get loaded twice:
1. bind
2. bind
3. destroy
4. Using the first engine <-- Many crashed happen here
4. destroy <-- Some crashes happen here

One such cases is from [11 years ago](https://openssl-dev.openssl.narkive.com/97zf5qT2/what-is-the-really-proper-way-to-use-an-engine) and is reproduced when loading the engine twice.

Recent change in the initialization flow around OpenSSL 1.1.0/1.1.1 moved the invocation of `CONF_modules_load_file()` causing multiple projects to invoke it twice, thus triggering engine errors such as [this one](https://bugs.launchpad.net/ubuntu/+source/openssl/+bug/1921518):
* wget - [bugfix](https://lists.gnu.org/archive/html/bug-wget/2019-04/msg00031.html)
* curl - [bugfix](https://bugs.launchpad.net/ubuntu/+source/curl/+bug/1940528)
* Issue #9481
* etc.

In fact, once Canonical and us debugged [this case](https://bugs.launchpad.net/ubuntu/+source/openssl/+bug/1921518) and found that both wget and curl crash due to loading the engine twice (PKA engine in our deployment), it led Canonical to start a package-wide search for other projects that might suffer from the same bug.

Suggestion:
-------------
While the different projects that use OpenSSL should fix their bug that cause them to load the configuration file twice, from the thread in #9481 it sounds like this shouldn't lead to segfaults in these programs, which is the case today. As loading an engine twice will always result in a failure to add it to the engine list, and the destruction of the just created engine, it seems like checking for the existence of an engine prior to creating / dynamically loading it, will resolve this repeating issue.

Again, it seems like errors / crashes caused by attempts to load an engine twice are a repeated issue around the OpenSSL API. I suggest to prevent the possibility of dynamically loading the engine if it was already loaded, and this will be a robust fix for this repeated issue.

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Use the address of the bind function as a DYNAMIC_ID,
since the true name of the engine is not known
before the bind function returns,
but invoking the bind function before the engine
is unloaded results in memory corruption.

Fixes #17023

This is an alternative to #17055

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
This looks much better than the alternative solution.

**Comment by beldmit** [comment]
I like this approach and kindly ask to add a test.

**Comment by bernd-edlinger** [comment]
okay, thanks for reminding me about the test case, one simple thing that was broken before this and gets fixed is
the memory leak I mentioned here:
https://github.com/openssl/openssl/pull/17025#discussion_r751385849
`../util/shlib_wrap.sh ./openssl s_client -engine $PWD/../engines/ossltest.so -engine $PWD/../engines/ossltest.so`
so it is a bit O/S dependent, though.

File: crypto/engine/eng_list.c [hunk 1]
```diff
@@ -278,6 +360,8 @@ static void engine_cpy(ENGINE *dest, const ENGINE *src)
     dest->load_pubkey = src->load_pubkey;
     dest->cmd_defns = src->cmd_defns;
     dest->flags = src->flags;
+    dest->dynamic_id = src->dynamic_id;
+    engine_add_dynamic_id(dest, NULL);
```
Comment (bernd-edlinger) [inline crypto/engine/eng_list.c:364]: oops, here the global lock is already held, so I will have to add a not_locked parameter to this function

**Review by t8m — APPROVED** [review]
Yeah adding a test case would be nice.

**Comment by beldmit** [comment]
I'd suggest to withdraw an approval for now.

**Comment by bernd-edlinger** [comment]
Test case added, please re-confirm.  Thanks!

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Good for all branches, I suppose cherry-pick to 1.1.1 will require separate PR.

**Review by beldmit — APPROVED** [review]
LGTM. Nice work!

File: crypto/engine/eng_list.c [hunk 2]
```diff
@@ -128,6 +134,84 @@ static int engine_list_remove(ENGINE *e)
...
+        }
+        if (e->dynamic_id != NULL)
+            goto err;
+        e->dynamic_id = dynamic_id;
+    }
+
+    if (engine_dyn_list_head == NULL) {
+        /* We are adding to an empty list. */
+        if (engine_dyn_list_tail != NULL)
+            return 0;
```
Comment (bernd-edlinger) [inline crypto/engine/eng_list.c:168]: oops, sorry, this should be a `goto err;`

File: crypto/engine/eng_list.c [hunk 3]
```diff
@@ -128,6 +134,84 @@ static int engine_list_remove(ENGINE *e)
...
+
+    if (engine_dyn_list_head == NULL) {
+        /* We are adding to an empty list. */
+        if (engine_dyn_list_tail != NULL)
+            goto err;
+        engine_dyn_list_head = e;
+        e->prev_dyn = NULL;
+    } else {
+        /* We are adding to the tail of an existing list. */
+        if ((engine_dyn_list_tail == NULL) || (engine_dyn_list_tail->next_dyn != NULL))
```
Comment (bernd-edlinger) [inline crypto/engine/eng_list.c:173]: this line is longer than 80 characters.
I will change that in a moment.

File: crypto/engine/eng_lib.c [hunk 4]
```diff
@@ -91,6 +92,7 @@ int engine_free_util(ENGINE *e, int not_locked)
     if (e->destroy)
         e->destroy(e);
     CRYPTO_free_ex_data(CRYPTO_EX_INDEX_ENGINE, e, &e->ex_data);
+    engine_remove_dynamic_id(e, not_locked);
```
Comment (bernd-edlinger) [inline crypto/engine/eng_lib.c:95]: hmm, in the debugger I see, that the DSO is unloaded in the CRYPTO_free_ex_data above,
so the dynamic_id should better be unlinked before that.
Thereforfe, I'm going to swap those two statements.

**Comment by bernd-edlinger** [comment]
Hi @t8m @beldmit
I think the PR is now finalized.
Please re-confirm your approvals.

Thanks
Bernd.

**Comment by petrovr** [comment]
I'm still not convinced that exist defect that needs code correction. 
Now I waste some time to test engines and result is https://github.com/openssl/openssl/issues/17023#issuecomment-974768758 .

File: crypto/engine/eng_dyn.c [hunk 5]
```diff
@@ -484,7 +484,9 @@ static int dynamic_load(ENGINE *e, dynamic_data_ctx *ctx)
     engine_set_all_null(e);
 
     /* Try to bind the ENGINE onto our own ENGINE structure */
-    if (!ctx->bind_engine(e, ctx->engine_id, &fns)) {
+    if (!engine_add_dynamic_id(e, (ENGINE_DYNAMIC_ID)ctx->bind_engine, 1)
+            || !ctx->bind_engine(e, ctx->engine_id, &fns)) {
+        engine_remove_dynamic_id(e, 1);
```
Comment (eitkin-nvidia) [inline crypto/engine/eng_dyn.c:489]: Wouldn't this mean that a double-loaded engine will also remove the dynamic_id of the original engine instance? I agree that the engine won't be loaded twice (bind_engine() won't be called), but if the load will happen 3 times won't the engine be loaded twice?

**Comment by eitkin-nvidia** [comment]
I agree, and prefer this suggested fix over my approach in #17055.

There is just the small issue I pointed to in the code, and once it is resolved I think it will be a god solution to the multi-load issue, thanks for taking the time to implement it on your side.

File: crypto/engine/eng_dyn.c (hunk 5 above)
Comment (eitkin-nvidia) [reply crypto/engine/eng_dyn.c:489]: nvm, I saw what you did the e->dynamic_id, and setting it up only after everything is good. Nice solution.

**Review by t8m — APPROVED** [review]

**Comment by bernd-edlinger** [comment]
Merged to all branches.  Thanks!

**Pull request closed without merging**